### PR TITLE
OSSL_PROVIDER_load_ex

### PR DESCRIPTION
**Pull request by beldmit**
to configure provider in load-time.
The tests are to be written, the documentation is also to be written.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
It's an attempt to implement #21350 function using available API.

@openssl/committers the comments wanted. I tried to implement as an extension of `OPENSSL_provider_load` but didn't find an easy option to provide configuration options.

File: crypto/provider_conf.c
Comment (mspncp) [inline crypto/provider_conf.c:361]: Without having gone into details too much yet:  the  name `OSSL_PROVIDER_activate()` seems to be a very misleading choice given the fact that

a) it is intended to be some alternative to `OSSL_provider_load()`
b) there already exists an unrelated function  `ossl_provider_activate()` where the name seems to be more fitting.

https://github.com/openssl/openssl/blob/e3d897d3fa3b48bb835fab0665a435469beea7ae/crypto/provider_core.c#L1274-L1292

How about one of the following names?
* `OPENSSL_provider_load_with_params()`
* `OPENSSL_provider_load_from_params()`

File: crypto/provider_conf.c
Comment (mattcaswell) [inline crypto/provider_conf.c:330]: There does not seem to be a way to add sections

File: crypto/provider_conf.c
Comment (beldmit) [reply crypto/provider_conf.c:361]: I thought about this or similar naming but current OSSL_PROVIDER_load returns a provider handle and doesn't give me any opportunity to add config options (at least I don't see them). But I don't insist on the name.

The implementation I provided is intended to load the provider and activate it.

File: crypto/provider_conf.c
Comment (mattcaswell) [inline crypto/provider_conf.c:351]: I find this implementation quite surprising. It is a way to create a config file from OSSL_PARAMs - but I understood the requirement to be to be able to configure providers from params. This takes params and converts them into a config file. Later that will be converted back into OSSL_PARAMs to pass to the provider. Why not just pass the OSSL_PARAMs straight to the provider without the config file middle-man.

File: crypto/provider_conf.c
Comment (mattcaswell) [inline crypto/provider_conf.c:348]: The name of this function is quite strange. It is only indirectly related to providers at all. It converts params to a config file and loads it. That might end up loading/configuring some providers or it might not depending on whats in the config file. It's even more indirectly related to activation (since it is perfectly possible to configure a provider without activating it).

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/provider_conf.c
Comment (beldmit) [reply crypto/provider_conf.c:351]: Yes. It's a quite ugly and hackerish (in bad sense) way to reach my goal.

I didn't find an internal API function passing params to provider -  `OSSL_PROVIDER_load` doesn't accept params and functions from `crypto/provider_conf.c` deal with parsed config, not params.

Comment (t8m) [reply crypto/provider_conf.c:351]: Instead of creating conf, you could create the sk_INFOPAIR stack of name/value items that the provider could query via the core_get_params() call. Or you could even modify the core_get_params() to return params directly but that would be more invasive change.

Comment (paulidale) [reply crypto/provider_conf.c:351]: What about adding get/set params calls at the provider level?

If we're using OSSL_PARAM as the passed type (& I believe we should since it was designed for this), let's just use it properly without any kludges.

Comment (t8m) [reply crypto/provider_conf.c:351]: That would require changing providers. IMO not worth it.

And we already do get params but as a callback from provider to core.

This proposal already does the OSSL_PROVIDER_load_ex() albeit with strange OSSL_PROVIDER_activate() name.

**Comment by beldmit** [comment]
I've pushed a draft of API design for better clarity

File: doc/prov_loadex.md
Comment (t8m) [inline doc/prov_loadex.md:71]: This is trying to solve a different problem than "providing the provider parameters from an application". I am not sure it is a good idea to mix up these two problems together. But yeah, I agree that we perhaps unnecessarily require holding a provider reference when an application just wants a provider to be loaded into a library context for its whole lifetime.

Perhaps the function should be named OSSL_LIB_CTX_load_provider() instead?

Now the question is what should happen if the provider is already loaded, i.e. the provider's initialization already happened with different parameters (or no parameters at all)?

**Review by t8m — COMMENTED** [review]
My feeling is this really needs some active discussion within the OTC to find out what we want to do here. We should avoid requiring changes to existing providers if they would like to use this functionality if at all possible. I assume some changes will be inevitable if we want to support setting the parameters to providers after they are loaded & initialized but at least for the initial load it should be possible to use the current callback.

File: doc/prov_loadex.md
Comment (beldmit) [reply doc/prov_loadex.md:53]: My intention here is the scenario: we have loaded the provider and specified most of the parameters (via config) but not all, and we want to specify the remaining ones.

File: doc/prov_loadex.md
Comment (Jakuje) [inline doc/prov_loadex.md:37]:
```suggestion
#Configure a provider in the config file, activate on demand
```

File: doc/prov_loadex.md
Comment (simo5) [reply doc/prov_loadex.md:57]: As @t8m commented, it will be too late.
Once the provider is initialized any change in params will be ignored.
We would have to add a provider API and change providers to provide these function to update parameters after initialization.

File: doc/prov_loadex.md
Comment (simo5) [reply doc/prov_loadex.md:76]: I think we may want to deinit the provider and reinit from scratch ?
Potentially dangerous though.

Alternatively we return an error (already loaded), and the application can take it from there.

Comment (beldmit) [reply doc/prov_loadex.md:76]: @simo5 I think that there may be parameters that can be set in arbitrary moment of time. But provider may reject all the params change otherwise than on initialization.

Comment (simo5) [reply doc/prov_loadex.md:76]: It is not how it works today, but you could explicitly propose behavior changes I guess?

**Comment by simo5** [comment]
> So having thought about this some more I think I agree with a OSSL_LIB_CTX_load_provider() call where the provider config is entirely provided via the params option.

If a provider is already loaded this call should fail, btw, up to the app to decide what to do with it.
 
> However I am entirely against OSSL_PROVIDER_set_params_by_name(), I think mixing static and dynamic config or attempting to change config will be error prone, difficult to debug for all involved (openssl, provider and application developers), and ultimately not a good pattern.

In general I think if the config is in a config file users should tweak the config file.
And if apps load the provider dynamically they should just set the params right at load time.

File: doc/prov_loadex.md
Comment (beldmit) [reply doc/prov_loadex.md:57]: OK, I got your point.

File: doc/prov_loadex.md
Comment (beldmit) [reply doc/prov_loadex.md:76]: Yes.

File: doc/prov_loadex.md
Comment (beldmit) [reply doc/prov_loadex.md]: Done, thanks!

**Comment by beldmit** [comment]
Thanks, added `OSSL_PROVIDER_settable_params` to the scenario

File: crypto/provider_conf.c
Comment (beldmit) [reply crypto/provider_conf.c:351]: `OSSL_PROVIDER_load` returns a provider handle and it would be strange if `OSSL_PROVIDER_load_ex` returns anything different. But I agree that the proposed name is bad.

Comment (paulidale) [reply crypto/provider_conf.c:351]: PARAMS were designed for the cross provider interface.  Using anything else means reinventing this or bodging something else in it's place.

We already use params to configure providers -- look at the options for the FIPS provider.

IMO, this is the only sensible way forwards.

Comment (beldmit) [reply crypto/provider_conf.c:351]: I agree but how do we pass them otherwise then in config form?

Comment (paulidale) [reply crypto/provider_conf.c:351]: I'm not sure what you mean here.
How to convert config to params keeping the nesting?
Or something else?

Comment (beldmit) [reply crypto/provider_conf.c:351]: We have an internal function (I extracted it myself :) ) that accepts a config structure and activates the provider. It's not clear to me how to pass the parameters directly to provider so that's why I dance this strange dance to use the maximum of the current code.

Comment (paulidale) [reply crypto/provider_conf.c:351]: We will need to discuss this offline I suspect ...

File: crypto/provider_conf.c
Comment (beldmit) [reply crypto/provider_conf.c:348]: I enforce activation in the built config. But yes, in general you are correct.

**Comment by beldmit** [comment]
OK, I realized how to pass the data to the provider in more or less standard way. 
Hopefully will push the tests tomorrow.

**Comment by beldmit** [comment]
This test failure seems to be unrelated to the patch but may be of interest for @hlandau https://github.com/openssl/openssl/actions/runs/5741052478/job/15560296780?pr=21604

**Comment by beldmit** [comment]
Not a draft anymore and can be normally reviewed.

**Comment by beldmit** [comment]
Close/reopen to work with the compilable master branch

**Comment by beldmit** [comment]
Ready to review

File: crypto/provider_core.c
Comment (mattcaswell) [inline crypto/provider_core.c:527]: Not sure why we are creating an "ex" version of an internal function. Just keep this as `ossl_provider_new` and convert all internal callers to the new param list.

File: crypto/provider_core.c
Comment (mattcaswell) [inline crypto/provider_core.c:584]: Doesn't this lead to a double free in the case where parameters are supplied from a config file? In that case we are doing a shallow copy of the `INFOPAIR` stack from the stack held in `store->provinfo`. Subsequently (when the store gets freed), `ossl_provider_store_free` gets called which also frees the `INFOPAIR` stack.

File: crypto/provider_core.c
Comment (mattcaswell) [inline crypto/provider_core.c:572]: So the parameters are additive to any parameters that have already been obtained from a config file?

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod:113]: This should describe what happens if parameters are supplied in both the config file *and* via these functions.

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod:111]: typo: accepting *an*

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod:235]: *The* .... functions *were* added

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/provider_core.c
Comment (beldmit) [reply crypto/provider_core.c:565]: I'm not sure. I think no because template is a local variable here and doesn't deal with the config file.

File: crypto/provider_core.c
Comment (beldmit) [reply crypto/provider_core.c:577]: Again, I think no - as template is a local variable.

File: doc/man3/OSSL_PROVIDER.pod
Comment (beldmit) [reply doc/man3/OSSL_PROVIDER.pod:113]: I believe that we don't get there from the config-provided code path though I am not sure

File: crypto/provider_conf.c
Comment (beldmit) [reply crypto/provider_conf.c]: Resolving this thread because I have implemented a normal way to deal with it

File: crypto/provider_conf.c
Comment (beldmit) [reply crypto/provider_conf.c]: Resolving this thread because the function is implemented in a normal way now.

File: doc/prov_loadex.md
Comment (beldmit) [reply doc/prov_loadex.md]: The file is removed from the PR, resolving the thread

File: crypto/provider_core.c
Comment (beldmit) [reply crypto/provider_core.c]: Done, thanks

File: doc/man3/OSSL_PROVIDER.pod
Comment (beldmit) [reply doc/man3/OSSL_PROVIDER.pod]: Fixed, thanks!

File: doc/man3/OSSL_PROVIDER.pod
Comment (beldmit) [reply doc/man3/OSSL_PROVIDER.pod]: Fixed, thanks!

File: crypto/provider_core.c
Comment (mattcaswell) [reply crypto/provider_core.c:565]: The `store->provinfo` code directly above checks the config information from the config file. In particular see the function `provider_conf_load` in the crypto/provider_conf.c file in the case that `activate == 0`: it prepopulates `store->provinfo` with config data for later use by `ossl_provider_new`.

See line 549 above which causes `template` to be populated with data from the config file.

The comment on line 544 is misleading since this is actually not just related to user added builtin providers.

File: crypto/provider_core.c
Comment (mattcaswell) [reply crypto/provider_core.c:577]: But line 549 above pre-populates it with data from `store->provinfo`

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [reply doc/man3/OSSL_PROVIDER.pod:113]: I believe that if the a provider is configured in the config file but *without* an "activate" line then the parameters supplied via `OSSL_PROVIDER_load_ex()` are *added* to those from the config file (in the way that you have currently written this).

File: crypto/provider_core.c
Comment (beldmit) [reply crypto/provider_core.c:570]: Thanks! Added the copying of the parameters.

File: crypto/provider_conf.c
Comment (beldmit) [reply crypto/provider_conf.c]: Not relevant anymore, resolving

File: crypto/provider_core.c
Comment (beldmit) [reply crypto/provider_core.c]: Fixed, thanks!

File: doc/man3/OSSL_PROVIDER.pod
Comment (beldmit) [reply doc/man3/OSSL_PROVIDER.pod]: Provided a clarification.

File: doc/man3/OSSL_PROVIDER.pod
Comment (mattcaswell) [inline doc/man3/OSSL_PROVIDER.pod:116]: I am wondering whether this "additive" behaviour is the correct one. If the parameters from config and the parameters from application code conflict in any way then the result could be unexpected.

I am not entirely sure what the right answer is here. Possible things we could do:

1) application params always take precedence over, and replace params in the config file
2) config file params always taken precedence over, and replace application params
3) application params are additive (as now)...but then what should happen if they both specify the same param? Or the params conflict in some non-obvious way?
4) Trying to specify application params where config params exist causes a failure
5) Some additional "flag" parameter which specifies what to do
6) Provide some mechanism for an application to query the config params??

Comment (beldmit) [reply doc/man3/OSSL_PROVIDER.pod]: The update is pushed.

**Review by mattcaswell — APPROVED** [review]

File: doc/man3/OSSL_PROVIDER.pod
Comment (simo5) [reply doc/man3/OSSL_PROVIDER.pod]: If you do 5 you need to consider multiple cases of potentially desired behavior:

a) only application provided parameters are considered
b) application parameters are merged to config provided, but applications parameters take precedence.
c) application parameters are merged to config provided, but config parameters take precedence.
d) only config parameters are used if available (ie the passed parameters are only a fallback if there is no config at all).

However if (6) is provided, then no flag needs to be implemented because all can be handled explicitly by the application in constructing the params array, and the behavior would be that the application always overrides the config file when explicitly loading params.
Helper functions to provide a,b,c,d behavior can then be provided outside of the loading mechanism, making the mechanism simpler.

File: crypto/provider_core.c
Comment (paulidale) [inline crypto/provider_core.c:567]: So non-UTF* string params are not allowed?

This isn't really passing PARAMs to the provider, I was kind of expecting a OSSL_PROVIDER_set_params() call implemented on the provider side.  OSSL_PARAM is our define mechanism for communicating across the libcrypto/provider boundary.

Comment (t8m) [reply crypto/provider_core.c:567]: IMO that would be a bad mechanism. We want to be able to configure existing providers this way and they would not have this set_params() call. Why wouldn't we use the existing mechanism of providers calling back to libcrypto?

Comment (t8m) [reply crypto/provider_core.c:567]: On the other hand we do not necessarily have to convert params passed by the application in the load_ex() calls to INFOPAIRs. So we could support any types of OSSL_PARAMS here. We could drop the INFOPAIR type altogether and just create UTF8 params from the config file as well.

Comment (beldmit) [reply crypto/provider_core.c:567]: Yes, limiting parameters to string type is intentional. 

I don't like the design here but I prefer to use the existing logic.

A separate problem is that when we explicitly pass the params to the application, we can't get them back.

Comment (paulidale) [reply crypto/provider_core.c:567]: > Why wouldn't we use the existing mechanism of providers calling back to libcrypto?

We're trying to push parameters into providers.  How would a provider know _when_ to call back to libcrypto for changed settings?

Startup/loading isn't the only time we might want to change things.

Comment (beldmit) [reply crypto/provider_core.c:567]: I think we don't plan to support change of the parameters just now

Comment (t8m) [reply crypto/provider_core.c:567]: > We're trying to push parameters into providers. How would a provider know _when_ to call back to libcrypto for changed settings?
> 
> Startup/loading isn't the only time we might want to change things.

Actually quite the opposite - we IMO do not want to change things after the provider initialization as that complicates provider implementation.

Or yeah, we might want to add the push-parameters-to-provider method eventually as well but that does not replace this proposal.

**Comment by beldmit** [comment]
The second review needed (@levitte would you mind?)

File: crypto/provider_core.c
Comment (levitte) [reply crypto/provider_core.c:567]: I don't really see a reason for a push model.  The provider code can retrieve whatever parameter it desires at any time it needs it, I don't see how that's a bad thing...

File: include/internal/provider.h
Comment (levitte) [inline include/internal/provider.h:35]: Nit: Please undo this incorrect reindentation

**Comment by beldmit** [comment]
Squashed and rebased for a normal review process

**Comment by beldmit** [comment]
Ready to review according the today's discussion.

File: doc/designs/prov_loadex.md
Comment (levitte) [inline doc/designs/prov_loadex.md:28]: The values from a configuration file are necessarily UTF8 strings because of its origin.  I'm not sure that's a reason why we *have* to limit parameters passed this way to UTF8 strings, though.  There is, of course, the caveat that non-UTF8 strings will not be possible to replicate into a config file...  question is, is that terrible?

Comment (beldmit) [reply doc/designs/prov_loadex.md:28]: I don't expect any problems while we have null-terminated strings here :)

**Comment by beldmit** [comment]
@openssl/committers any review/feedback, please?

**Comment by beldmit** [comment]
Rebased to resolve the trivial conflict in util/libcrypto.num

**Review by t-j-h — APPROVED** [review]

**Comment by beldmit** [comment]
Merged. Thanks for the review!

**Review by mspncp — COMMENTED** [review]
Hi @beldmit, sorry once again for my late review, I didn't manage to get it done earlier. Hope you find something useful in it nevertheless.

File: doc/designs/prov_loadex.md
Comment (levitte) [reply doc/designs/prov_loadex.md:28]: I frankly don't see it as a *restriction*, but fair enough.  I think it is important to have some level of expectation on humanly readable strings, and UTF8 was a choice that I based on how universal that encoding of non-ASCII chars has become.

**Comment by beldmit** [comment]
@mspncp could you please raise your documentation updates as a PR?

File: doc/designs/prov_loadex.md
Comment (mspncp) [reply doc/designs/prov_loadex.md:9]: done

File: doc/designs/prov_loadex.md
Comment (mspncp) [reply doc/designs/prov_loadex.md:32]: done

File: doc/designs/prov_loadex.md
Comment (mspncp) [reply doc/designs/prov_loadex.md:40]: done

File: doc/designs/prov_loadex.md
Comment (mspncp) [reply doc/designs/prov_loadex.md:63]: Added, but is this indeed what you meant to say? (! You may answer this question on the PR, TBD)

**Comment by mspncp** [comment]
See PR #22029. Please also reconsider the conversations which I did not mark as resolved yet.

File: doc/designs/prov_loadex.md
Comment (mspncp) [reply doc/designs/prov_loadex.md:54]: _(You may answer this question in PR #22029)_

File: doc/designs/prov_loadex.md
Comment (beldmit) [reply doc/designs/prov_loadex.md:28]: Exactly. The current internal representation is C string (and UTF8 works as C string) so it just a simple solution.

File: doc/designs/prov_loadex.md
Comment (beldmit) [reply doc/designs/prov_loadex.md:24]: Thanks!

File: test/provider_test.c
Comment (beldmit) [reply test/provider_test.c:175]: The parameter is not unknown, but I understand your reasons and yes, a FIPS provider test would be better.

File: doc/designs/prov_loadex.md
Comment (beldmit) [reply doc/designs/prov_loadex.md:12]: I agree though I totally missed this point

File: doc/designs/prov_loadex.md
Comment (beldmit) [reply doc/designs/prov_loadex.md:54]: Yes, exactly, thanks for the clarification!

**Pull request closed without merging**
